simulate socket timeout

brettporter opened the pull request:
By passing in a socket timeout to the nock request, it will emit a timeout event on the request if setTimeout has been called with a lower value. It does not actually wait for the specified period of time.

I've checked that this works with the base http library's request.setTimeout and socket.setTimeout, both with a "once" function and an "on timeout" handler.

I wanted to first check if this approach makes sense, and if so I can flesh it out with tests and documentation.

This is related to #164, though for one specific case.

brettporter commented:
Note: the emit error that is failing the tests is one thing that needs to be rethought - the core libraries don't actually emit an error, but they destroy the socket which will result in an error being emitted. Would appreciate your thoughts on how to handle that appropriately.

pgte commented on lib/socket.js:
```diff
@@ -10,7 +10,21 @@ function Socket() {
+    this.timeout = timeout;
+    this.timeoutFunction = fn;
+  }
+  socket.checkTimeout = function(delay) {
```
I believe this is an internal function, perhaps it should be prefixed with `_`?

pgte commented:
Other than this, it's looking good!
(Look out for the failing tests).

pgte commented:
ha, I missed your comment on the failing test. Let me check this first.

pgte commented:
Regarding the abortion, I think it shouldn't emit an error.
I think that the appropriate behaviour is similar to the way the real request.abort is implemented:

https://github.com/iojs/io.js/blob/v1.x/lib/_http_client.js#L168-L194

brettporter commented:
Thanks for looking over it!

> I think it shouldn't emit an error.
I think that the appropriate behaviour is similar to the way the real request.abort is implemented:
> 
> https://github.com/iojs/io.js/blob/v1.x/lib/_http_client.js#L168-L194

I agree. I've added some changes which instead of emitting an error on abort, it will do it if you attempt to write, end or respond to the request instead - this should match how a normal request behaves if a socket is destroyed (as is the case in the code above).

Does that make sense?

pgte commented:
Yes it does!

brettporter commented:
Hi @pgte - I've added tests and documentation. Also, writing the tests made it clear that calling it `socketTimeout` was confusing, as you're simulating a delay / idle socket (that may or may not lead to a timeout). I've renamed it to `socketDelay` which is hopefully clearer and a bit closer to the other `delay*` methods.

Is there anything else you need?

pgte commented:
@brettporter Thanks!

landed on v1.2.0
